CMakeLists.txt

```diff
@@ -0,0 +1,270 @@
+set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wstrict-prototypes -Wpointer-arith -Wsystem-headers)
+
+if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
+  message(STATUS "Setting build type to 'Release' as none was specified.")
+  set(CMAKE_BUILD_TYPE "Release")
+endif()
+
+set(CMAKE_C_FLAGS_DEBUG "-g3")
+set(CMAKE_C_FLAGS_RELEASE "-O3 -funroll-loops -fomit-frame-pointer")
+set(CMAKE_C_FLAGS_RELWITHDEBINFO "-g3 -O2")
+set(CMAKE_C_FLAGS_MINSIZEREL "-Os")
+
+if(COMPILE_LTO)
+    set(LTM_C_FLAGS  ${LTM_C_FLAGS} -flto)
+    set(LTM_LD_FLAGS ${LTM_LD_FLAGS} -flto)
+endif()
+
+# What compiler do we have and what are their...uhm... peculiarities
+if(CMAKE_C_COMPILER_ID MATCHES "(C|c?)lang")
+    set(LTM_C_FLAGS  ${LTM_C_FLAGS} -Wno-typedef-redefinition -Wno-tautological-compare -Wno-builtin-requires-header)
+    # Clang requires at least '-O1' for dead code eliminiation
+    set(CMAKE_C_FLAGS_DEBUG  "${CMAKE_C_FLAGS_DEBUG} -O1")
+endif()
+if(CMAKE_C_COMPILER MATCHES "mingw")
+    set(LTM_C_FLAGS  ${LTM_C_FLAGS}  -Wno-shadow)
+endif()
+if(CMAKE_SYSTEM_NAME MATCHES "Darwin")
+    set(LTM_C_FLAGS  ${LTM_C_FLAGS}  -Wno-nullability-completeness)
+endif()
+if(CMAKE_SYSTEM_NAME MATCHES "CYGWIN")
+    set(LTM_C_FLAGS  ${LTM_C_FLAGS} -no-undefined)
+endif()
+
+# TODO: coverage (lgcov)
+
+# If the user set the environment variables at generate-time, append them
+# in order to allow overriding our defaults.
+# ${LTM_CFLAGS} means the user passed it via sth like:
+# $ cmake -DLTM_CFLAGS="foo"
+set(LTM_C_FLAGS  ${LTM_C_FLAGS} ${LTM_CFLAGS})
+set(LTM_LD_FLAGS  ${LTM_LD_FLAGS} ${LTM_LDFLAGS})
+
+#-----------------------------------------------------------------------------
+# library target
+#-----------------------------------------------------------------------------
+add_library(${PROJECT_NAME}
+    ${SOURCES}
+)
+
+target_include_directories(${PROJECT_NAME} PUBLIC
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
+    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME}>
+)
+
+target_compile_options(${PROJECT_NAME} BEFORE PRIVATE
+    ${LTM_C_FLAGS}
+)
+target_link_options(${PROJECT_NAME} BEFORE PRIVATE
+    ${LTM_LD_FLAGS}
+)
+
+set_target_properties(${PROJECT_NAME} PROPERTIES
+    VERSION ${PROJECT_VERSION}
+    SOVERSION ${PROJECT_VERSION_MAJOR}
+)
+
+#-----------------------------------------------------------------------------
+# demo target
+#-----------------------------------------------------------------------------
+
+
+add_executable(test-target EXCLUDE_FROM_ALL
+    ${CMAKE_CURRENT_SOURCE_DIR}/demo/shared.c
+    ${CMAKE_CURRENT_SOURCE_DIR}/demo/test.c
+)
+
+target_include_directories(test-target PRIVATE
+    ${CMAKE_CURRENT_SOURCE_DIR}
+)
+
+target_link_libraries(test-target PRIVATE
+    ${PROJECT_NAME}
+)
+
+target_compile_options(test-target BEFORE PRIVATE
+    $<$<STREQUAL:$<TARGET_PROPERTY:${PROJECT_NAME},TYPE>,SHARED_LIBRARY>:-O1 -DLTM_TEST_DYNAMIC>
```

Comment:
> Lowering the optimization level to -O1 might hide bugs caused by additional optimizations.

Anybody remember what the exact reasons were for this restriction?

Reply:
> Why do you override the optimization level here and below?
People should be able build/run the tests in debug/relase/... mode.

In the versions before we didn't use the CMake specific modes, but only `LTM_C_FLAGS` and so this `-O1` was overridden by the user when `LTM_C_FLAGS` was appended.

Now we use those CMake specific variables and there's no way to prepend something ...

> There is target_compile_definitions to add compile definitions.

this doesn't help, since we still have to make sure to have at least `-O1`